java/ql/lib/ext/org.apache.commons.httpclient.model.yml

```diff
@@ -0,0 +1,7 @@
+extensions:
+  - addsTo:
+      pack: codeql/java-all
+      extensible: sinkModel
+    data:
+      - ["org.apache.commons.httpclient", "URI", True, "URI", "(String,boolean)", "", "Argument[0]", "open-url", "generated"]
```

Comment:
should we use "ai-generated"?

Reply:
Also, should this `URI` model potentially be a step to be consistent with how both [`java.net.URI#URI(String)`](https://github.com/github/codeql/blob/71c8bb20f96b616dbdb76e31bd6858831fd73792/java/ql/lib/ext/java.net.model.yml#L24) and [`java.net.URL#URL(String)`](https://github.com/github/codeql/blob/71c8bb20f96b616dbdb76e31bd6858831fd73792/java/ql/lib/ext/java.net.model.yml#L29) are currently modeled as taint steps? Or is the intention to have this particular case as a sink instead? Just want to confirm 🙂

Reply:
I don't think we can use arbitrary values in the `provenance` column, but I'll defer to @michaelnebel to confirm that.

> Also, should this URI model potentially be a step to be consistent with how both [java.net.URI#URI(String)](https://github.com/github/codeql/blob/71c8bb20f96b616dbdb76e31bd6858831fd73792/java/ql/lib/ext/java.net.model.yml#L24) and [java.net.URL#URL(String)](https://github.com/github/codeql/blob/71c8bb20f96b616dbdb76e31bd6858831fd73792/java/ql/lib/ext/java.net.model.yml#L29) are currently modeled as taint steps?

Yes, we agreed with @kaeluka that we were happy for these models to include some things that may need manual correction (like this one), as long as they were highlighted as important enough to take a look at. Deciding whether to add them as sinks or change them to steps is an expected part of the review process.

Reply:
I have now changed them to `ai-generated`. If `@michaelnebel` tells us to do something else, I'll change it again.

> e.g. Should ATM-generated models be included in the data that the [metrics query](https://github.com/github/codeql/blob/d3e73891b55f3c7a70fa583dfcf74909570f4e76/java/ql/src/Metrics/Summaries/GeneratedVsManualCoverage.ql#L4) for generated versus manual MaD coverage considers, or should the ATM data-points be kept separate?

It'd be GREAT if the ATM models could have their own provenance!

Reply:
I think it is a good idea for ATM models to have their own provenance and `ai-generated` sounds like a good descriptive name.
It will require that we
(1) Update some of our unit tests that checks the provenance column
(2) Go through part of our dataflow library code to make that models wil `ai-generated` is actually being consumed. Without further investigation I think that will automatically work for sources and sinks, but it will not be the case for summaries - for those we do explicit ranking and *inclusion* based on their provenance.
Do you expect to add summary models as well?
Maybe I push some changes to this PR as well?

Reply:
Thanks, @michaelnebel!

> Do you expect to add summary models as well?

Not for now, but I'll expect this to come up at *some* point.

> Maybe I push some changes to this PR as well?

I'd be very grateful if you did <3 Thank you!

Reply:
Great!
The `ai-generated` is now included as a valid provenance value, but this will only work as expected for sinks and sources.
Are you intending to include summaries as well in this PR?

Reply:
I don't intend to include summaries in this PR 👍 Thank you :)

Reply:
With the latest changes that should also work - but only for Java.
Are you also generating models for C#? :-)

Reply:
No, currently only for Java :)

Reply:
> Also, should this URI model potentially be a step to be consistent with how both [java.net.URI#URI(String)](https://github.com/github/codeql/blob/71c8bb20f96b616dbdb76e31bd6858831fd73792/java/ql/lib/ext/java.net.model.yml#L24)

I've now removed this from the PR for the time being so we might be able to merge this soon. I run into cases like this regularly, so we will definitely need to have this discussion :)

Reply:
Thanks @jcogs33 for raising the following important questions in [this comment](https://github.com/github/codeql/pull/12228#discussion_r1109981958):

> Should there potentially be a different provenance for ATM-generated models? The "generated" provenance currently means that the model was created by the other model generator tool. Will distinguishing between ATM-generated and models created by the model generator tool matter?
>
> e.g. Should ATM-generated models be included in the data that the metrics query for generated versus manual MaD coverage considers, or should the ATM data-points be kept separate?

Our current meaning of the provenance column is as follows:
- A "manual" model has been manually verified by a human (regardless of how it was found or generated), and we have very high confidence in that data point being correct.
- A "generated" model was bulk-generated and not verified manually (unless a manual model also exists).

We want to maintain that differentiation in order to continue to be able to measure the accuracy of our generated approaches. At the same time, it seems useful to store provenance information on the mechanism through which a model was originally found, even if it was subsequently verified manually.

One possibility: store the generating mechanism as a prefix `ai-`, `hq-` (heuristic query), `df-` (data flow library) followed by a suffix of `generated` (not verified manually) or `manual` (manually verified, even if originally generated by the method in the prefix). The prefix could be optional, only used if the origin is known.

For this PR, if the above approach is used, the models would then have provenance `ai-manual` since they are manually verified after being generated.

Thoughts? What code would need to be modified to take this into account and ensure our metrics and other code continues to work? Any alternative suggestions? @michaelnebel @atorralba

Reply:
From a CodeML perspective, we're happy with any solution that lets us count the number of ai-generated declarations that were merged into the Java modeling. So `ai-manual` is 💯

Reply:
Good point @yo-h. I think the format `(generatingMechanism)-(verifiedOrNot)` could work great for all purposes.

> What code would need to be modified to take this into account and ensure our metrics and other code continues to work?

I guess this needs some refactor in our code related to metrics, MaD (external flow/flow summaries), and the model generator. @michaelnebel probably will know the specifics better. Maybe we should do that separately in another PR, to not block/scope creep this one?

For the time being, we could backtrack and merge this as `manual` models, until we do the proper refactor.

Reply:
Yes, that sounds like a good solution. We could definitely do something like that. It will require some re-write of the framework code and model generator code, but it shouldn't be too much trouble.
I pushed some commits in this PR for supporting `ai-generated` models, so it is still an option to go with that provenance for the models to be merged in this PR. Then the provenance can be changed to `ai-manual`, when the framework and model generator code has been re-written to support the new provenance format.

Reply:
Is that the approach we should take for provenance kinds (then I will put it on my todo list)?

Reply:
> I pushed some commits in this PR for supporting ai-generated models, so it is still an option to go with that provenance for the models to be merged in this PR. Then the provenance can be changed to ai-manual, when the framework and model generator code has been re-written to support the new provenance format.

This makes it easier to change all of them to `ai-manual` later. I'm happy to change them all to `manual` now, but they'll be harder to find later on :) Should we go with this suggestion?